File: CSharp/ModulusTreeCryptography/ModulusTreeCryptography/ModulusFileDecryption.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace ModulusTreeCryptography
{
    public static class ModulusFileDecryption
    {
        private const int chunkSize = 17;

        public static byte[] DecryptBytes(ModulusTreeKey key, byte[] inputBytes)
        {
            var chunkCount = inputBytes.Length / chunkSize;
            var extraBytes = inputBytes.Length % chunkSize;

            var values = new List<BigInteger>();

            for (int i = 0; i < chunkCount; i++)
            {
                var offset = i * chunkSize;
                var chunk = inputBytes.Skip(offset).Take(chunkSize).ToArray();

                var output = DecryptChunk(key, chunk);
                values.Add(output);
            }

            if (extraBytes != 0)
            {
                var chunk = inputBytes.Skip(chunkCount * chunkSize).Take(extraBytes).ToArray();

                var output = DecryptChunk(key, chunk);
                values.Add(output);
            }

            return values.SelectMany(x =>
                                     {
                                         var bytes = x.ToByteArray();
                                         var last = bytes.Last();
                                         if (last == 0 || last == 1) //TODO: Potential bug! How to decide when to chop off the sign bit?
                                             bytes = bytes.Take(bytes.Length - 1).ToArray();
                                         return bytes;
                                     }).ToArray();
        }

        private static BigInteger DecryptChunk(ModulusTreeKey key, byte[] chunk)
        {
            var input = new BigInteger(chunk);
            var output = ModulusDecryption.GetValueAtIndex(key, input);
            return output;
        }
    }
}

File: Java/src/edu/byui/cs480/modulustree/ModulusFileDecryption.java
package edu.byui.cs480.modulustree;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

/**
 * Created by clay on 5/17/14.
 */
public class ModulusFileDecryption
{
    private static final int CHUNK_SIZE = 17;

    public static Byte[] decryptBytes(ModulusTreeKey pKey, Byte[] pInputBytes)
    {
        int chunkCount = pInputBytes.length / CHUNK_SIZE;
        int extraBytes = pInputBytes.length % CHUNK_SIZE;

        List<BigInteger> values = new ArrayList<BigInteger>();

        for (int i = 0; i < chunkCount; i++)
        {
            int offset = i * CHUNK_SIZE;
            Byte[] chunk = Stream.of(pInputBytes).skip(offset).limit(CHUNK_SIZE).toArray(Byte[]::new);
            BigInteger output = decryptChunk(pKey, chunk);
            values.add(output);
        }
        if (extraBytes != 0)
        {
            Byte[] chunk = Stream.of(pInputBytes).skip(chunkCount * CHUNK_SIZE).limit(extraBytes).toArray(Byte[]::new);
            BigInteger output = decryptChunk(pKey, chunk);
            values.add(output);
        }

        ArrayList<Byte> retVal = new ArrayList<>();
        for (BigInteger x : values)
        {
            byte[] bytes = x.toByteArray();
            int requiredPadding = 16 - bytes.length;
            for (int i = 0; i < requiredPadding; i++)
                retVal.add((byte)0);
            for (byte b : bytes)
                retVal.add(b);
        }
        int padding = retVal.remove(0);
        Byte[] ret = new Byte[retVal.size() - padding];
        retVal.subList(0, retVal.size() - padding).toArray(ret);
        return ret;
    }

    private static BigInteger decryptChunk(ModulusTreeKey pKey, Byte[] pChunk)
    {
        //Unbox
        int i = 0;
        byte[] bytes = new byte[pChunk.length];
        for (Byte b : pChunk)
        {
            bytes[i++] = b;
        }
        BigInteger input = new BigInteger(bytes);
        return ModulusDecryption.getValueAtIndex(pKey, input);
    }
}
